src/components/ripple/ripple.ts

```diff
@@ -0,0 +1,288 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  HostBinding,
+  Input,
+  OnChanges,
+  OnDestroy,
+  OnInit,
+  Renderer,
+  SimpleChange,
+  ViewEncapsulation,
+} from '@angular/core';
+
+enum RippleState {
+  NEW,
+  EXPANDING,
+  FADING_OUT,
+}
+
+class Ripple {
+  state = RippleState.NEW;
+  constructor(public element: Element) {}
+}
+
+const RIPPLE_SPEED_PX_PER_SECOND = 500;
+const MIN_RIPPLE_FILL_TIME_SECONDS = 0.2;
+const MAX_RIPPLE_FILL_TIME_SECONDS = 0.6;
+
+const sqr = (x: number) => x * x;
+
+const distanceToFurthestCorner = (x: number, y: number, parentRect: any) => {
+  const x1 = parentRect.left;
+  const x2 = x1 + parentRect.width;
+  const y1 = parentRect.top;
+  const y2 = y1 + parentRect.height;
+  const maxSquaredDistance = Math.max(
+      sqr(x - x1) + sqr(y - y1),
+      sqr(x2 - x) + sqr(y - y1),
+      sqr(x - x1) + sqr(y2 - y),
+      sqr(x2 - x) + sqr(y2 - y));
```

Comment:
Using rect.top/left/right/bottom without temp variables, which I think makes it self-explanatory.